So far, we have created our Mocks but we have simply defined what the returned value should be.

What if we want to have a test where we test for an exception. So, we want our mock to throw an exception of some specific type.

For this, we can use the "thenThrow" method instead of "thenReturn".

And that's it!

    @Test
    @DisplayName("Throws an Exception")
    public void testThrowException() {

        // Since it is a prototype bean, we will get a new instance each time we request it
        CollegeStudent nullStudent = context.getBean("collegeStudent", CollegeStudent.class);

        when(applicationDao.checkNull(nullStudent)).thenThrow(new RuntimeException());

        assertThrows(RuntimeException.class, () -> applicationService.checkNull(nullStudent));
    }

What if we want to make consecutive calls where for first call, it throws an exception but for second call, it returns something? Well, we can attach "thenThrow" and "thenReturn" one after the other. And they will act like two calls.

    @Test
    @DisplayName("Throws an Exception with two calls")
    public void testThrowExceptionTwoCalls() {

        // Since it is a prototype bean, we will get a new instance each time we request it
        CollegeStudent nullStudent = context.getBean("collegeStudent", CollegeStudent.class);

        when(applicationDao.checkNull(nullStudent))
                .thenThrow(new RuntimeException())
                .thenReturn("Do not throw exception the second time");

        assertThrows(RuntimeException.class, () -> applicationService.checkNull(nullStudent));

        assertEquals("Do not throw exception the second time", applicationService.checkNull(nullStudent));
    }

The order matters here.

If you write the "assertEquals" before "assertThrows", then it will do this "assertEquals" check for the first method call which actually throws an exception so the test will fail.

But, if you do not write the "assertEquals" altogether and only have "assertThrows", that won't affect the test because "assertThrows" will always test for the first call.